### ISSUE TITLE
active internet connection required when starting docker instance of Kokoro-TTS

### ISSUE THREAD

**Issue by gitchat1** [remsky/Kokoro-FastAPI#194]
I finally managed to build the GPU docker image and noticed that it requires an internet connection whem starting up. At least when it is done via docker desktop.  Is there any way to change this?

**Comment by fireblade2534** [comment]
The changes that were merged to main 8 hours ago should have allowed you to start it without internet

**Comment by fireblade2534** [comment]
right now you will have to rebuild (just pull the latest version of the repo, and the run your build command. do not clear your docker build cache as docker will automatically detect what needs to be rerun)

**Comment by gitchat1** [comment]
Thanks I noticed that for whatwver reason docker didn't recognize my graphics card anymore and for some reason I also didn't have internet connectivity anymore so I switched to Ubuntu. I wasn't able to built the image there probably because I couldn't configure rootless mode for docker using the Nvidia toolkit. The prebuilt does work now however. Any idea when the prebuilt will be able to completely run offline?

**Comment by Sakuradancing** [comment]
@fireblade2534 
Hello, I am looking for a solution that can run offline on an arm machine and use the CPU for text-to-speech generation. Kokoro is a very suitable work in all aspects. When I use docker to run kokoro-fastapi-cpu:v0.2.2, it seems that it needs to download some resources from github before it can run. When I checked this issue, I noticed that the error situation is similar to mine. Also，when I checked the released version, I found that the version was released quite quickly more than ten days ago, but there has been no new version released recently. Do you have any plans to update the docker image in the near future? Hope to get a reply, thank you!

My device version is Ubuntu 20.04 aarch64 offline
![Image](https://github.com/user-attachments/assets/6188ef8b-1d1b-4385-8a04-5ad9d266bd3f)

**Comment by gitchat1** [comment]
Okay I managed to build the programme using this yaml file.

name: kokoro-tts-gpu
services:
  kokoro-tts:
    build:
      context: ../..
      dockerfile: docker/gpu/Dockerfile
    volumes:
      - api_volume:/app/api
    ports:
      - "8880:8880"
    environment:
      - PYTHONPATH=/app
      - USE_GPU=true
      - PYTHONUNBUFFERED=1
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: all
              capabilities: [gpu]

volumes:
  api_volume:
    name: kokoro-tts-api

I apperently still need an internet connection

> ERROR [kokoro-tts internal] load metadata for docker.io/nvidia/cuda:1  0.0s
------
 > [kokoro-tts internal] load metadata for docker.io/nvidia/cuda:12.8.0-cudnn-runtime-ubuntu24.04:
------
failed to solve: nvidia/cuda:12.8.0-cudnn-runtime-ubuntu24.04: failed to resolve source metadata for docker.io/nvidia/cuda:12.8.0-cudnn-runtime-ubuntu24.04: failed to authorize: failed to fetch anonymous token: Get "https://auth.docker.io/token?scope=repository%3Anvidia%2Fcuda%3Apull&service=registry.docker.io": dial tcp: lookup auth.docker.io on 127.0.0.53:53: server misbehaving

**Comment by fireblade2534** [comment]
@Sakuradancing and @gitchat1 
Do these exact commands:
git clone https://github.com/remsky/Kokoro-FastAPI.git
cd Kokoro-FastAPI/docker/gpu
sudo docker compose up --build

Now u can stop the container and disconnect from the internet. To start the container do (you can add -d if u want it to continue without needing you to leave your terminal open):
sudo docker compose up

**Comment by gitchat1** [comment]
Thanks for carring!

First off. the yaml-file included with the project usually failed to build for me. This did the trick under Mint.

`name: kokoro-tts-gpu
services:
  kokoro-tts:
    build:
      context: ../..
      dockerfile: docker/gpu/Dockerfile
    volumes:
      - api_volume:/app/api
    ports:
      - "8880:8880"
    environment:
      - PYTHONPATH=/app
      - USE_GPU=true
      - PYTHONUNBUFFERED=1
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: all
              capabilities: [gpu]

volumes:
  api_volume:
    name: kokoro-tts-api`

I followed your instructions and got the following log.

`$ git clone https://github.com/remsky/Kokoro-FastAPI.git
cd Kokoro-FastAPI/docker/gpu
sudo docker compose up --build
fatal: Zielpfad 'Kokoro-FastAPI' existiert bereits und ist kein leeres Verzeichnis.
Compose now can delegate build to bake for better performances
Just set COMPOSE_BAKE=true
[+] Building 1.8s (18/18) FINISHED                               docker:default
 => [kokoro-tts internal] load build definition from Dockerfile            0.0s
 => => transferring dockerfile: 1.97kB                                     0.0s
 => [kokoro-tts internal] load metadata for docker.io/nvidia/cuda:12.8.0-  1.6s
 => [kokoro-tts internal] load .dockerignore                               0.0s
 => => transferring context: 367B                                          0.0s
 => [kokoro-tts stage-0  1/12] FROM docker.io/nvidia/cuda:12.8.0-cudnn-ru  0.0s
 => [kokoro-tts internal] load build context                               0.0s
 => => transferring context: 7.75kB                                        0.0s
 => CACHED [kokoro-tts stage-0  2/12] RUN apt-get update && apt-get insta  0.0s
 => CACHED [kokoro-tts stage-0  3/12] RUN curl -LsSf https://astral.sh/uv  0.0s
 => CACHED [kokoro-tts stage-0  4/12] RUN useradd -m -u 1001 appuser &&    0.0s
 => CACHED [kokoro-tts stage-0  5/12] WORKDIR /app                         0.0s
 => CACHED [kokoro-tts stage-0  6/12] COPY --chown=appuser:appuser pyproj  0.0s
 => CACHED [kokoro-tts stage-0  7/12] RUN --mount=type=cache,target=/root  0.0s
 => CACHED [kokoro-tts stage-0  8/12] COPY --chown=appuser:appuser api ./  0.0s
 => CACHED [kokoro-tts stage-0  9/12] COPY --chown=appuser:appuser web ./  0.0s
 => CACHED [kokoro-tts stage-0 10/12] COPY --chown=appuser:appuser docker  0.0s
 => CACHED [kokoro-tts stage-0 11/12] RUN chmod +x ./entrypoint.sh         0.0s
 => CACHED [kokoro-tts stage-0 12/12] RUN if [ "true" = "true" ]; then     0.0s
 => [kokoro-tts] exporting to image                                        0.0s
 => => exporting layers                                                    0.0s
 => => writing image sha256:e0d0db17b4b23d07f8623c4b1cdcdf402c7d586669ddd  0.0s
 => => naming to docker.io/library/kokoro-tts-gpu-kokoro-tts               0.0s
 => [kokoro-tts] resolving provenance for metadata file                    0.0s
[+] Running 2/2
 ✔ kokoro-tts                             Built                            0.0s
 ✔ Container kokoro-tts-gpu-kokoro-tts-1  Created                          0.0s
Attaching to kokoro-tts-1
kokoro-tts-1  |
kokoro-tts-1  | ==========
kokoro-tts-1  | == CUDA ==
kokoro-tts-1  | ==========
kokoro-tts-1  |
kokoro-tts-1  | CUDA Version 12.8.0
kokoro-tts-1  |
kokoro-tts-1  | Container image Copyright (c) 2016-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
kokoro-tts-1  |
kokoro-tts-1  | This container image and its contents are governed by the NVIDIA Deep Learning Container License.
kokoro-tts-1  | By pulling and using the container, you accept the terms and conditions of this license:
kokoro-tts-1  | https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license
kokoro-tts-1  |
kokoro-tts-1  | A copy of this license is made available in this container at /NGC-DL-CONTAINER-LICENSE for your convenience.
kokoro-tts-1  |
kokoro-tts-1  | 2025-03-02 19:21:12.122 | INFO     | __main__:download_model:60 - Model files already exist and are valid
kokoro-tts-1  | INFO:     Started server process [30]
kokoro-tts-1  | INFO:     Waiting for application startup.
kokoro-tts-1  | 07:21:17 PM | INFO     | main:57 | Loading TTS model and voice packs...
kokoro-tts-1  | 07:21:17 PM | INFO     | model_manager:38 | Initializing Kokoro V1 on cuda
kokoro-tts-1  | 07:21:17 PM | DEBUG    | paths:101 | Searching for model in path: /app/api/src/models
kokoro-tts-1  | 07:21:17 PM | INFO     | kokoro_v1:45 | Loading Kokoro model on cuda
kokoro-tts-1  | 07:21:17 PM | INFO     | kokoro_v1:46 | Config path: /app/api/src/models/v1_0/config.json
kokoro-tts-1  | 07:21:17 PM | INFO     | kokoro_v1:47 | Model path: /app/api/src/models/v1_0/kokoro-v1_0.pth
kokoro-tts-1  | /app/.venv/lib/python3.10/site-packages/torch/nn/modules/rnn.py:123: UserWarning: dropout option adds dropout after all but last recurrent layer, so non-zero dropout expects num_layers greater than 1, but got dropout=0.2 and num_layers=1
kokoro-tts-1  |   warnings.warn(
kokoro-tts-1  | /app/.venv/lib/python3.10/site-packages/torch/nn/utils/weight_norm.py:143: FutureWarning: `torch.nn.utils.weight_norm` is deprecated in favor of `torch.nn.utils.parametrizations.weight_norm`.
kokoro-tts-1  |   WeightNorm.apply(module, name, dim)
kokoro-tts-1  | 07:21:18 PM | DEBUG    | paths:153 | Scanning for voices in path: /app/api/src/voices/v1_0
kokoro-tts-1  | 07:21:18 PM | DEBUG    | paths:131 | Searching for voice in path: /app/api/src/voices/v1_0
kokoro-tts-1  | 07:21:18 PM | DEBUG    | model_manager:77 | Using default voice 'af_heart' for warmup
kokoro-tts-1  | 07:21:18 PM | INFO     | kokoro_v1:73 | Creating new pipeline for language code: a
kokoro-tts-1  | 07:21:19 PM | DEBUG    | kokoro_v1:245 | Generating audio for text with lang_code 'a': 'Warmup text for initialization.'
kokoro-tts-1  | 07:21:19 PM | DEBUG    | kokoro_v1:252 | Got audio chunk with shape: torch.Size([57600])
kokoro-tts-1  | 07:21:19 PM | INFO     | model_manager:84 | Warmup completed in 2035ms
kokoro-tts-1  | 07:21:19 PM | INFO     | main:101 |
kokoro-tts-1  |
kokoro-tts-1  | ░░░░░░░░░░░░░░░░░░░░░░░░
kokoro-tts-1  |
kokoro-tts-1  |     ╔═╗┌─┐┌─┐┌┬┐
kokoro-tts-1  |     ╠╣ ├─┤└─┐ │
kokoro-tts-1  |     ╚  ┴ ┴└─┘ ┴
kokoro-tts-1  |     ╦╔═┌─┐┬┌─┌─┐
kokoro-tts-1  |     ╠╩╗│ │├┴┐│ │
kokoro-tts-1  |     ╩ ╩└─┘┴ ┴└─┘
kokoro-tts-1  |
kokoro-tts-1  | ░░░░░░░░░░░░░░░░░░░░░░░░
kokoro-tts-1  |                 
kokoro-tts-1  | Model warmed up on cuda: kokoro_v1CUDA: True
kokoro-tts-1  | 67 voice packs loaded
kokoro-tts-1  |
kokoro-tts-1  | Beta Web Player: http://0.0.0.0:8880/web/
kokoro-tts-1  | or http://localhost:8880/web/
kokoro-tts-1  | ░░░░░░░░░░░░░░░░░░░░░░░░
kokoro-tts-1  |
kokoro-tts-1  | INFO:     Application startup complete.
kokoro-tts-1  | INFO:     Uvicorn running on http://0.0.0.0:8880 (Press CTRL+C to quit)


w Enable Watch

`
The problem is it does not seem to have built anything as you can see when looking at this output.

`REPOSITORY                  TAG       IMAGE ID       CREATED        SIZE
zonos-zonos                 latest    561b4d45e6a9   30 hours ago   16.9GB
kokoro-tts-gpu-kokoro-tts   latest    e0d0db17b4b2   4 days ago     13.1GB

`
All I can think of is deleting rge image and starting from scratch. Not sure if that would help though.

**Comment by fireblade2534** [comment]
that looks fine to me?

**Comment by fireblade2534** [comment]
if it works Id say its fine. Also what does the default docker build command produce? (Like with the default compose file)

**Comment by gitchat1** [comment]
1. Shut down the container.
2. Disconnect from the internet and bring the container online
3. Then without reconnecting to the internet visit http://localhost:8880/web
If our images are the same you won't be able to enter any text or select any voice via the web interface. When you are online durring the initial phase the problem does not occur. The same is true if you reconnect to the internet and reload the page.

As for what error the original file produced? I think it might be nest if you read through this thread here. https://github.com/remsky/Kokoro-FastAPI/issues/195

**Comment by Sakuradancing** [comment]
@fireblade2534

**Comment by fireblade2534** [comment]
Honestly I don't have an ARM CPU and don't know much about ARM so i'm not really sure. I did compare the CPU's with passmark https://www.cpubenchmark.net/compare/4756vs4677/ARM-Cortex-A55-4-Core-2124-MHz-vs-Intel-i5-12400 (Do note that the ARM-Cortex-A55 has a high margin for error as they only have one sample) it does appear that it does not have even close to the performance though this is also likely a result of software not having as good support for ARM compared to
 x86.

**Comment by Sakuradancing** [comment]
@fireblade2534 
Hello, thanks for your reply. From the website you provided, we can see that there is a huge performance gap between the two CPUs (10-20 times). My guess is that the gap is in frequency, but I didn't expect that there is also a huge gap in other aspects. I would like to express my gratitude for your help, and I hope that our communication can help other people in similar situations. Then I personally make an immature speculative summary for future generations:
1. If your usage environment is offline, only CPU, arm version, it is recommended to use docker for deployment, and use kokoro-fastapi-cpu:v0.2.3pre and later versions
2. The CPU level will greatly affect the speed of speech synthesis, just like i5 12400 and A55, the synthesis speed of the two is also between 10-20 times the theoretical speed (reference https://www.cpubenchmark.net/compare/4756vs4677/ARM-Cortex-A55-4-Core-2124-MHz-vs-Intel-i5-12400)

**Comment by remsky** [comment]
The docker images should be fully self contained with everything baked in now. I notice this seemed to split into other issues, feel free to reopen if any occur on a recent release
